tstest/tlstest/gencerts.go

```diff
@@ -0,0 +1,173 @@
+			Organization: []string{"Tailscale Test Org"},
+		},
+		NotBefore: now,
+		NotAfter:  now.AddDate(10, 0, 0),
+
+		IsCA:                  true,
+		BasicConstraintsValid: true,
+		KeyUsage:              x509.KeyUsageCertSign | x509.KeyUsageCRLSign,
+		SubjectKeyId:          mustSKID(&key.PublicKey),
+	}
+
+	der, err := x509.CreateCertificate(rand.Reader, tpl, tpl, &key.PublicKey, key)
+	if err != nil {
+		return fmt.Errorf("create root: %w", err)
+	}
+
+	if err := writePEM("root-ca.crt", "CERTIFICATE", der); err != nil {
+		return err
+	}
+	if err := marshalECKey("root-ca.key", key); err != nil {
+		return err
+	}
+	log.Printf("wrote %s and %s", "root-ca.crt", "root-ca.key")
+	return nil
+}
+
+// generateLeafCert signs a new cert for host using the existing CA files.
+func generateLeafCert(host string) error {
+	caCert, err := readCert("root-ca.crt")
+	if err != nil {
+		return err
+	}
+	caPriv, err := readECKey("root-ca.key")
+	if err != nil {
+		return err
+	}
+
+	leafKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if err != nil {
+		return err
+	}
+
+	serial, err := rand.Int(rand.Reader, big.NewInt(0).Lsh(big.NewInt(1), 128))
+	if err != nil {
+		return err
+	}
+
+	now := time.Now().Add(-time.Hour)
+	tpl := &x509.Certificate{
+		SerialNumber: serial,
+		Subject:      pkix.Name{CommonName: host},
+		NotBefore:    now,
+		NotAfter:     now.AddDate(2, 0, 0),
+
+		KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageKeyEncipherment,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		BasicConstraintsValid: true,
+		DNSNames:              []string{host},
+	}
+
+	der, err := x509.CreateCertificate(rand.Reader, tpl, caCert, &leafKey.PublicKey, caPriv)
+	if err != nil {
+		return err
+	}
+
+	crtPath := host + ".crt"
+	keyPath := host + ".key"
+	if err := writePEM(crtPath, "CERTIFICATE", der); err != nil {
+		return err
+	}
+	if err := marshalECKey(keyPath, leafKey); err != nil {
+		return err
+	}
+
+	log.Printf("wrote %s and %s", crtPath, keyPath)
+	return nil
+}
+
+func writePEM(path, typ string, der []byte) error {
+	f, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0o600)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	return pem.Encode(f, &pem.Block{Type: typ, Bytes: der})
+}
+
+func marshalECKey(path string, key *ecdsa.PrivateKey) error {
+	der, err := x509.MarshalECPrivateKey(key)
+	if err != nil {
+		return err
+	}
+	return writePEM(path, "EC PRIVATE KEY", der)
+}
+
+func readCert(path string) (*x509.Certificate, error) {
+	pemBytes, err := os.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+	block, _ := pem.Decode(pemBytes)
+	if block == nil || block.Type != "CERTIFICATE" {
+		return nil, fmt.Errorf("no CERTIFICATE in %s", path)
+	}
+	return x509.ParseCertificate(block.Bytes)
+}
+
+func readECKey(path string) (*ecdsa.PrivateKey, error) {
+	pemBytes, err := os.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+	block, _ := pem.Decode(pemBytes)
```

Comment:
nit: check for trailing data, that should be an error too

Reply:
done